feat: implements AWS signature version 4 for signing requests

bojeil-google opened the pull request:
This is based on:
https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html

It will be used to generate signed requests to AWS GetCallerIdentity API.
https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html

This API is used to return details about the IAM user or role whose
credentials are used to call the operation.

grayside committed:
refactor(samples): idtoken-cloudrun => idtoken-serverless (#1025)

yoshi-automation committed:
docs: README updated with new sample paths

release-please[bot] committed:
chore: release 6.0.7 (#1023)

* chore: updated samples/package.json [ci skip]

* chore: updated CHANGELOG.md [ci skip]

* chore: updated package.json

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

yoshi-automation committed:
build: credential-file-override is no longer required (#1029)

This PR was generated using Autosynth. :rainbow:

Synth log will be available here:
https://source.cloud.google.com/results/invocations/5f7f9c6d-c75a-4c60-8bb8-0026a14cead7/targets

- [ ] To automatically regenerate this PR, check this box.

Source-Link: https://github.com/googleapis/synthtool/commit/94421c47802f56a44c320257b2b4c190dc7d6b68

yoshi-automation committed:
chore: update cloud rad kokoro build job (#1034)

This PR was generated using Autosynth. :rainbow:

Synth log will be available here:
https://source.cloud.google.com/results/invocations/b742586e-df31-4aac-8092-78288e9ea8e7/targets

- [ ] To automatically regenerate this PR, check this box.

Source-Link: https://github.com/googleapis/synthtool/commit/bd0deaa1113b588d70449535ab9cbf0f2bd0e72f

renovate-bot committed:
fix(deps): roll back dependency google-auth-library to ^6.0.6 (#1033)

SurferJeffAtGoogle committed:
build: perform publish using Node 12 (#1035)

This PR was generated using Autosynth. :rainbow:



- [ ] To automatically regenerate this PR, check this box.

Source-Link: https://github.com/googleapis/synthtool/commit/5747555f7620113d9a2078a48f4c047a99d31b3e

release-please[bot] committed:
chore: release 6.0.8 (#1036)

* chore: updated samples/package.json [ci skip]

* chore: updated CHANGELOG.md [ci skip]

* chore: updated package.json

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

xethlyx committed:
docs: correct spelling of its (#1040)

Minor change that fixes a spelling error. I don't think an issue needed to be opened for this. Thanks!

banyan committed:
docs: fix usage in readme (#1031)

Add await

Co-authored-by: Justin Beckwith <justin.beckwith@gmail.com>
Co-authored-by: sofisl <55454395+sofisl@users.noreply.github.com>
Co-authored-by: Benjamin E. Coe <bencoe@google.com>

yoshi-automation committed:
chore: start tracking obsolete files (#1043)

This PR was generated using Autosynth. :rainbow:

Synth log will be available here:
https://source.cloud.google.com/results/invocations/7a1b0b96-8ddb-4836-a1a2-d2f73b7e6ffe/targets

- [ ] To automatically regenerate this PR, check this box.

yoshi-automation committed:
build: move system and samples test from Node 10 to Node 12 (#1045)

This PR was generated using Autosynth. :rainbow:

Synth log will be available here:
https://source.cloud.google.com/results/invocations/ba2d388f-b3b2-4ad7-a163-0c6b4d86894f/targets

- [ ] To automatically regenerate this PR, check this box.

Source-Link: https://github.com/googleapis/synthtool/commit/05de3e1e14a0b07eab8b474e669164dbd31f81fb

bojeil-google committed:
Merge remote-tracking branch 'upstream/byoid'

bojeil-google committed:
feat: implements AWS signature version 4 for signing requests

This is based on:
https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html

It will be used to generate signed requests to AWS GetCallerIdentity API.
https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html

This API is used to return details about the IAM user or role whose
credentials are used to call the operation.

bcoe commented on src/auth/awsrequestsigner.ts:
```diff
@@ -0,0 +1,280 @@
+import {Headers} from './oauth2client';
+import {Crypto, createCrypto, fromArrayBufferToHex} from '../crypto/crypto';
+
+type HttpMethod =
```
Curious if we use a similar type anywhere else in the codebase, or if `gaxios` exposes it? If not, no objection to having it here.

bojeil-google replied:
It seems that `gaxios` [does not export](https://github.com/googleapis/gaxios/blob/5b586d52b1798b6edf2bec7e59402e4d4115c8ee/src/common.ts#L69) a dedicated type for this.

I couldn't find any similar type in the current repo.

bcoe commented on src/auth/awsrequestsigner.ts:
```diff
@@ -0,0 +1,280 @@
+  const dateStamp = now.toISOString().replace(/[-]/g, '').replace(/T.*/, '');
+
+  // Change all additional headers to be lower case.
+  const reformattedAdditionalAmzHeaders: Headers = {};
```
This logic is definitely the most complex I think I've bumped into during this review, I believe OAuth 1.0 worked similarly, with regards to  needing to sort and sign parameters?

bojeil-google replied:
I completely agree. The crypto they use is not trivial and it is not standard. You are right on the similarities with OAuth 1.0 but at least the latter is a standard and there is sufficient documentation on it. 
We had to follow the [AWS documentation](https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html) in order to generate these (step by step). They do provide reference input/outputs for each step which helped. 

I think the bulk of the logic is documented in `generateAuthenticationHeaderMap`. I provided links to the 4 steps and some comments to clarify intermediate steps. I hope this makes it easier for contributors to follow. One possibility that was floated around in the early stages was to pull in AWS SDK as a dependency. However, I think it would be too expensive for serverless and low memory environments like GCF, IOT devices, etc. The auth library is a dependency for a lot of other libraries where this feature is not used so it could be very taxing to these users.

bcoe commented on test/test.awsrequestsigner.ts:
```diff
@@ -0,0 +1,297 @@
+    // List of various requests and their expected signatures.
+    // Examples source:
+    // https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html
+    const getRequestOptionsTests: AwsRequestSignerTest[] = [
```
I like this DSL you've built out for exercising the the fairly complex logic in the AWS request signer.

bojeil-google replied:
Thank you. I think this simplifies adding and removing new tests for this logic. It also makes it easier to replace the existing canonical request/signatures we are using. So it should not be expensive to maintain and build on in the future.

bcoe reviewed (approved):
Reading through this, I definitely am left wishing there was less complexity 😆 

But I also think the implementation is sound, and doesn't introduce any unnecessary complexity.

bojeil-google committed:
Adds additional AWS signature v4 tests from AWS botocore tests.
https://github.com/boto/botocore/tree/879f8440a4e9ace5d3cf145ce8b3d5e5ffb892ef/tests/unit/auth/aws4_testsuite

Adds support for raw date header in AWS requests.

Fixes padding issue in `fromArrayBufferToHex`.

bcoe commented on test/test.awsrequestsigner.ts:
```diff
@@ -100,15 +98,446 @@ describe('AwsRequestSigner', () => {
+            headers: {
+              Authorization:
+                'AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20110909/us-east-1/host/' +
+                `aws4_request, SignedHeaders=date;host;zoo, Signature=${signature}`,
```
thank you for pulling these tests over from boto 👍 sorry for being a pain.
